elastic/kibana
# Add origin conflict checks when resolving import errors

Resolves #124392.

Before, origin conflicts were only checked when initially importing a saved object. Subsequent calls to resolve import errors (a different API) skipped the origin conflict check. If you ran into a "missing references" error -- which are currently only triggered if your object has a reference to an data view (index pattern) or a search object -- that would prevent Kibana from checking for origin conflicts for the object, and a subsequent call to resolve the missing references error would wind up creating a duplicate instead of conflicting with an object of the same origin.

Now, resolving import errors also checks for origin conflicts, but it skips those checks if the consumer has specified a different destination for the retry, or if they are using the `createNewCopies: true` option ("Create new objects with random IDs" in the UI).

### Testing

(Copied steps 1-9 from the linked issue)

1. Start Kibana with a fresh ES cluster
2. Go to **Home**, then add the "Sample eCommerce orders" sample data
3. Go to **Stack Management** -> **Spaces**, then create "Another" space
4. Go to **Stack Management** -> **Saved Objects**, find the "[eCommerce] Orders by Country" map object, and export it with the "Include related objects" option _disabled._ This map has a reference to a data view and will not be functional without it
   - Attached here for convenience: [export.ndjson.zip](https://github.com/elastic/kibana/files/8017631/export.ndjson.zip)
5. Go to Another space
6. Go to **Stack Management** -> **Data Views**, then create a new data view using the existing sample data index, you can name it "`kibana_sample*`". This is the same Elasticsearch data index that the map relies on in the Default space, but this data view has a different object ID, it is not the same object
7. Go to **Stack Management** -> **Saved Objects** and import the file using the default options ("Check for existing objects" and "Automatically overwrite conflicts")
8. You will see this callout, pick the drop-down and select your new data view object:
   ![image](https://user-images.githubusercontent.com/5295965/152845588-7ae0fdb0-3f68-4667-b094-f2c59fc3aaaa.png)
9. Repeat steps 7-8
10. Observe that the object with a matching origin has been overwritten, and Kibana did not create a duplicate:
   ![image](https://user-images.githubusercontent.com/5295965/153424329-c5aa41fd-36e9-4a0b-aaf1-60752b29124d.png)

The import/resolve flow also works as expected when other options are selected:

* "Check for existing objects" with "Request action on conflict" -> prompts the user to overwrite the object after fixing the missing reference error
* "Create new objects with random IDs" -> creates a new object with an empty origin

### Technical details

When you initially use the option to "Automatically overwrite conflicts", that passes the `overwrite: true` option to the import API. But the resolve import errors API doesn't have such an option, instead it expects overwrites to be individually selected for retries.

The client is smart enough to remember that the user initially selected this, though, and doesn't surface a popup telling them the object needs to be overwritten. Instead, it makes another request to resolve import errors _again_, this time telling Kibana how to resolve the missing references error _and_ the conflict error.

In steps 9-10 above, three API calls are made:

1. `POST /api/saved_objects/_import?overwrite=true`
   Request body:
   ```js
   {
     file: (binary)
   }
   ```
   Response: "missing_references" error
2. `POST /api/saved_objects/_resolve_import_errors`
   Request body:
   ```js
   {
     file: (binary),
     retries: [{
       id: '2c9c1f60-1909-11e9-919b-ffe5949a18d2',
       type: 'map',
       replaceReferences: [{
         type: 'index-pattern',
         from: 'ff959d40-b880-11e8-a6d9-e546fe2bba5f',
         to: '0a512b70-8847-11ec-8cbd-d3d7e69811b9'
       }]
     }]
   ```
   Response: "conflict" error, `destinationId: '2c9c1f60-1909-11e9-919b-ffe5949a18d2'`
3. `POST /api/saved_objects/_resolve_import_errors`
   Request body:
   ```js
   {
     file: (binary),
     retries: [{
       id: '2c9c1f60-1909-11e9-919b-ffe5949a18d2',
       type: 'map',
       overwrite: true,
       destinationId: 'dd64a1d2-c429-4ed1-b33c-657cd07c465b',
       replaceReferences: [{
         type: 'index-pattern',
         from: 'ff959d40-b880-11e8-a6d9-e546fe2bba5f',
         to: '0a512b70-8847-11ec-8cbd-d3d7e69811b9'
       }]
     }]
   ```
   Response: OK :+1:

-----

**Release note:** When importing a saved object, resolving an "Index pattern conflict" error will no longer create a duplicate saved object.
